fe/src/main/java/org/apache/doris/load/routineload/RoutineLoad.java

```diff
@@ -0,0 +1,200 @@
+        try {
+            return beIdTomaxConcurrentTasks.values().stream().mapToInt(i -> i).sum();
+        } finally {
+            readUnlock();
+        }
+    }
+
+    public void addRoutineLoadJob(RoutineLoadJob routineLoadJob) {
+        writeLock();
+        try {
+            idToRoutineLoadJob.put(routineLoadJob.getId(), routineLoadJob);
+        } finally {
+            writeUnlock();
+        }
+    }
+
+    public void addRoutineLoadTasks(List<RoutineLoadTask> routineLoadTaskList) {
+        writeLock();
+        try {
+            idToRoutineLoadTask.putAll(routineLoadTaskList.parallelStream().collect(
+                    Collectors.toMap(task -> task.getSignature(), task -> task)));
+        } finally {
+            writeUnlock();
+        }
+    }
+
+    public Map<Long, RoutineLoadTask> getIdToRoutineLoadTask() {
+        return idToRoutineLoadTask;
+    }
+
+    public void addNeedSchedulerRoutineLoadTasks(List<RoutineLoadTask> routineLoadTaskList) {
+        writeLock();
+        try {
+            idToNeedSchedulerRoutineLoadTask.putAll(routineLoadTaskList.parallelStream().collect(
+                    Collectors.toMap(task -> task.getSignature(), task -> task)));
+        } finally {
+            writeUnlock();
+        }
+    }
+
+    public Map<Long, RoutineLoadTask> getIdToNeedSchedulerRoutineLoadTasks() {
+        readLock();
+        try {
+            return idToNeedSchedulerRoutineLoadTask;
+        } finally {
+            readUnlock();
+        }
+    }
+
+    public List<RoutineLoadJob> getRoutineLoadJobByState(RoutineLoadJob.JobState jobState) throws LoadException {
+        List<RoutineLoadJob> jobs = new ArrayList<>();
+        Collection<RoutineLoadJob> stateJobs = null;
+        readLock();
+        LOG.debug("begin to get routine load job by state {}", jobState.name());
+        try {
+            switch (jobState) {
+                case NEED_SCHEDULER:
+                    stateJobs = idToNeedSchedulerRoutineLoadJob.values();
+                    break;
+                case PAUSED:
+                    throw new LoadException("not support getting paused routine load jobs");
+                case RUNNING:
+                    stateJobs = idToRunningRoutineLoadJob.values();
+                    break;
+                case STOPPED:
+                    throw new LoadException("not support getting stopped routine load jobs");
+                default:
+                    break;
+            }
+            if (stateJobs != null) {
+                jobs.addAll(stateJobs);
+                LOG.info("got {} routine load jobs by state {}", jobs.size(), jobState.name());
+            }
+        } finally {
+            readUnlock();
+        }
+        return jobs;
+    }
+
+    public void updateRoutineLoadJobState(RoutineLoadJob routineLoadJob, RoutineLoadJob.JobState jobState) {
+        writeLock();
+        try {
+            RoutineLoadJob.JobState srcJobState = routineLoadJob.getState();
```

Comment:
Maybe check validation of state transfer will be implement in another place such as mysql statement validation.

Reply:
I think the best place to check this is here.
Because you check it somewhere else, e.g., analyze phase, you can not make it an atomic operation.
The check and transfer of state should be in one locking block.